File: generated-sdks/typescript/src/streaming.ts
/**
 * SSE streaming helpers for the Argentor TypeScript SDK.
 */

/**
 * Parse a Server-Sent Events response into an async generator of parsed data
 * objects. Stops when it encounters a `data: [DONE]` message.
 */
export async function* parseSSEStream(
  response: Response,
): AsyncGenerator<Record<string, unknown>> {
  const reader = response.body?.getReader();
  if (!reader) {
    throw new Error('Response body is not readable');
  }

  const decoder = new TextDecoder();
  let buffer = '';

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() ?? '';

      for (const line of lines) {
        const trimmed = line.trim();
        if (trimmed.startsWith('data: ')) {
          const data = trimmed.slice('data: '.length).trim();
          if (data === '[DONE]') {
            return;
          }
          yield JSON.parse(data) as Record<string, unknown>;
        }
      }
    }

    // Process any remaining buffer content
    if (buffer.trim().startsWith('data: ')) {
      const data = buffer.trim().slice('data: '.length).trim();
      if (data && data !== '[DONE]') {
        yield JSON.parse(data) as Record<string, unknown>;
      }
    }
  } finally {
    reader.releaseLock();
  }
}

File: generated-sdks/python/argentor_client/streaming.py
"""argentor_client — SSE streaming helpers."""

import json
from typing import Iterator, AsyncIterator, Any


class SSEStream:
    """Parses a Server-Sent Events stream from raw lines."""

    def __init__(self, lines: Iterator[str]):
        self._lines = lines

    def __iter__(self) -> Iterator[dict]:
        for line in self._lines:
            if line.startswith("data: "):
                data = line[len("data: "):]
                stripped = data.strip()
                if stripped == "[DONE]":
                    return
                yield json.loads(stripped)


class AsyncSSEStream:
    """Parses a Server-Sent Events stream from async lines."""

    def __init__(self, lines: AsyncIterator[str]):
        self._lines = lines

    async def __aiter__(self) -> AsyncIterator[dict]:
        async for line in self._lines:
            if line.startswith("data: "):
                data = line[len("data: "):]
                stripped = data.strip()
                if stripped == "[DONE]":
                    return
                yield json.loads(stripped)
